osu.Game/Database/LegacyArchiveExporter.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using osu.Framework.Platform;
+using osu.Game.Extensions;
+using osu.Game.Overlays.Notifications;
+using Realms;
+using SharpCompress.Common;
+using SharpCompress.Writers;
+using SharpCompress.Writers.Zip;
+
+namespace osu.Game.Database
+{
+    public abstract class LegacyArchiveExporter<TModel> : LegacyModelExporter<TModel>
+        where TModel : RealmObject, IHasNamedFiles, IHasGuidPrimaryKey
+    {
+        protected LegacyArchiveExporter(Storage storage, RealmAccess realm)
+            : base(storage, realm)
+        {
+        }
+
+        protected override void ExportToStream(TModel model, Stream outputStream, ProgressNotification notification, CancellationToken cancellationToken = default)
+            => exportZipArchive(model, outputStream, notification, cancellationToken);
+
+        /// <summary>
+        /// Exports an item to Stream as a legacy (.zip based) package.
+        /// </summary>
+        /// <param name="model">The model will be exported.</param>
+        /// <param name="outputStream">The output stream to export to.</param>
+        /// <param name="notification">The notification will displayed to the user</param>
+        /// <param name="cancellationToken">The Cancellation token that can cancel the exporting.</param>
+        private void exportZipArchive(TModel model, Stream outputStream, ProgressNotification notification, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                using var writer = new ZipWriter(outputStream, new ZipWriterOptions(CompressionType.Deflate));
+
+                float i = 0;
+                bool fileMissing = false;
+
+                foreach (var file in model.Files)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    using (var stream = UserFileStorage.GetStream(file.File.GetStoragePath()))
+                    {
+                        // Sometimes we cannot find the file(probably deleted by the user), so we handle this and post a error.
+                        if (stream == null)
+                        {
+                            // Only pop up once to prevent spam.
+                            if (!fileMissing)
+                            {
+                                PostNotification?.Invoke(new SimpleErrorNotification
+                                {
+                                    Text = "Some of your files are missing, they will not be included in the archive"
```

Comment:
What on earth? This is shocking. What is this logic even? Why is it here? And when is it ever tolerable to tell a user that "some of their files are missing"?

Reply:
This was an accidental case for me, maybe because I imported a skin that used the same file and deleted one, or some other accident, the realm couldn't find the skin file.
If this exception occurs, the entire package will not be exported in old logic, but this is not fatal, so I choose to pop up the prompt and continue

Reply:
I don't yet know how we should handle this kind of thing. I can imagine a scenario where this does happen, and I agree we probably want to present this to the user in some way, but I'm not 100% sure doing it with a notification like this is correct.

Similar cases could come up, for instance, when starting gameplay and files are missing. I'd hope we could standardise the way these are presented across the game. So probably wouldn't be against removing this localised handling for now (or at least replacing with a loud `Logger.Log`).